qutebrowser/javascript/webengine_caret.js

```diff
@@ -0,0 +1,1401 @@
+     * of endCursor and the next non-whitespace character.
+     * @param {Cursor} startCursor On exit, points to the position before
+     *     the char.
+     * @param {Cursor} endCursor The position to start searching for the next
+     *     char.  On exit, will point to the position past the char.
+     * @param {Array<Node>} nodesCrossed Any HTML nodes crossed between the
+     *     initial and final cursor position will be pushed onto this array.
+     * @param {boolean} skipWhitespace If true, will keep scanning until a
+     *     non-whitespace character is found.
+     * @return {?string} The next char, or null if the bottom of the
+     *     document has been reached.
+     */
+    TraverseUtil.getNextChar = function(
+        startCursor, endCursor, nodesCrossed, skipWhitespace) {
+        // Save the starting position and get the first character.
+        startCursor.copyFrom(endCursor);
+        let fChar = TraverseUtil.forwardsChar(endCursor, nodesCrossed);
+        if (fChar === null) {
+            return null;
+        }
+
+        // Keep track of whether the first character was whitespace.
+        const initialWhitespace = TraverseUtil.isWhitespace(fChar);
+
+        // Keep scanning until we find a non-whitespace or non-skipped character.
+        while ((TraverseUtil.isWhitespace(fChar)) ||
+            (TraverseUtil.isSkipped(endCursor.node))) {
+            fChar = TraverseUtil.forwardsChar(endCursor, nodesCrossed);
+            if (fChar === null) {
+                return null;
+            }
+        }
+        if (skipWhitespace || !initialWhitespace) {
+            // If skipWhitepace is true, or if the first character we encountered
+            // was not whitespace, return that non-whitespace character.
+            startCursor.copyFrom(endCursor);
+            startCursor.index--;
+            return fChar;
+        }
+
+        for (let i = 0; i < nodesCrossed.length; i++) {
+            if (TraverseUtil.isSkipped(nodesCrossed[i])) {
+                // We need to make sure that startCursor and endCursor aren't
+                // surrounding a skippable node.
+                endCursor.index--;
+                startCursor.copyFrom(endCursor);
+                startCursor.index--;
+                return " ";
+            }
+        }
+        // Otherwise, return all of the whitespace before that last character.
+        endCursor.index--;
+        return " ";
+    };
+
+    /**
+     * The class handling the Caret Browsing implementation in the page.
+     * Sets up communication with the background page, and then when caret
+     * browsing is enabled, response to various key events to move the caret
+     * or selection within the text content of the document. Uses the native
+     * Chrome selection wherever possible, but displays its own flashing
+     * caret using a DIV because there's no native caret available.
+     * @constructor
+     */
```

Comment:
You might want to rewrite this comment as it isn't really accurate anymore in our case. Or you can probably just remove it, this one doesn't seem very useful.